## Cleaning List of ICD10 Diagnoses with Psychotic Features

The stakeholder had given me a list of qualifying diagnoses, which I converted into a txt file to read into Python. 

In [1]:
import re
import pandas as pd

# import the txt file with the qualifying ICD10 diagnoses
with open('icd10dx.txt', 'r') as dx:
    dx_read = dx.read()

# put all the lines into the list
dx_list = dx_read.split("\n")
len(dx_list)

258

In [4]:
# check the list
dx_list[10:20]

['F23 Brief Psychotic Disorder',
 'F24 Shared Psychotic Disorder',
 'F25.0 Schizoaffective Disorder, Bipolar Type',
 'F25.1 Schizoaffective Disorder, Depressive Type',
 'F25.8 Other Schizoaffective Disorders',
 'F25.9 Schizoaffective Disorder, Unspecified',
 'F28 Other Psychotic Disorder Not Due to a Substance or',
 'Known Physiological Condition',
 'F29 Unspecified Psychosis Not Due to a Substance or',
 'Known Physiological Condition']

At first glance, it appears that there are 258 diagnoses in this file, but upon closer examination of what is in our list, we can see that there are some formatting errors. 

Most diagnoses are a string with an alpha-numeric code (ICD10) followed by a description:

> 'F25.0 Schizoaffective Disorder, Bipolar Type'

However, it seems like some diagnoses with a longer description got split into two lines:

> 'F28 Other Psychotic Disorder Not Due to a Substance or'

> 'Known Physiological Condition'

Thus, this must be accounted for. We can do this by leveraging regex. Dropping these lines from the description is something we can afford since matching on ICD10 code is much more important. We can query the diagnosis description again in the SQL database.

In [5]:
# get the filtered list of lines with dx only
dx_filtered = []

for line in dx_list:
    if re.match(r"[A-Z]\d", line):
        dx_filtered.append(line)

print(len(dx_filtered))

212


After this, we need to separate the ICD10 code from the description and put this into a Pandas dataframe

In [8]:
# separate the dx codes and the dx_desc
dx_codes = []
dx_desc = []

for line in dx_filtered:
    dx_codes.append(line.split(" ")[0])
    dx_desc.append(" ".join(line.split(" ")[1:]))
    
# put into dataframe and output as csv for ease of use
dx_df = pd.DataFrame(dx_desc, index = dx_codes)
dx_df.head()
#dx_df.to_csv('dx_filtered_file.csv')

0
F20.0          Paranoid Schizophrenia
F20.1      Disorganized Schizophrenia
F20.2         Catatonic Schizophrenia
F20.3  Undifferentiated Schizophrenia
F20.5          Residual Schizophrenia

I can also leverage Python to write some of SQL script for me, particularly the temp table to hold these diagnoses.

In [11]:
for dx in dx_codes:
    print("INSERT INTO #dx values ('{}')".format(dx))

INSERT INTO #dx values ('F20.0')
INSERT INTO #dx values ('F20.1')
INSERT INTO #dx values ('F20.2')
INSERT INTO #dx values ('F20.3')
INSERT INTO #dx values ('F20.5')
INSERT INTO #dx values ('F20.81')
INSERT INTO #dx values ('F20.89')
INSERT INTO #dx values ('F20.9')
INSERT INTO #dx values ('F21')
INSERT INTO #dx values ('F22')
INSERT INTO #dx values ('F23')
INSERT INTO #dx values ('F24')
INSERT INTO #dx values ('F25.0')
INSERT INTO #dx values ('F25.1')
INSERT INTO #dx values ('F25.8')
INSERT INTO #dx values ('F25.9')
INSERT INTO #dx values ('F28')
INSERT INTO #dx values ('F29')
INSERT INTO #dx values ('F30.10')
INSERT INTO #dx values ('F30.11')
INSERT INTO #dx values ('F30.12')
INSERT INTO #dx values ('F30.13')
INSERT INTO #dx values ('F30.2')
INSERT INTO #dx values ('F30.3')
INSERT INTO #dx values ('F30.4')
INSERT INTO #dx values ('F30.8')
INSERT INTO #dx values ('F30.9')
INSERT INTO #dx values ('F31.0')
INSERT INTO #dx values ('F31.10')
INSERT INTO #dx values ('F31.11')
INSERT INTO #d